File: app/MMarket/js/src/pgs/acc/Profits.js
// Copyright 11-Jun-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

/**
    Accounting profits.
**/

import Domo from "../../dmjs/Domo.js"; //eslint-disable-line
import Ui from "../../dmjs/Ui.js";
import {Menu} from "../../dmjs/Menu.js";
import {_} from "../../I18n.js";
import Cts from "../../data/Cts.js";
import ProfitsWg from "./wgs/ProfitsWg.js";

const $ = e => Ui.$(e);

/**
    Accounting profits.
**/
export default class Profits {
  /**
      @param {!Domo} wg
      @param {!Array<?>} data Has the following structure:
        [
          [
            date:string [total:number, acc:number, risk:number],
            date [total, acc, risk],
            ... (from after to before)
            until N dates
          ]
          [
            date [total, acc, risk],
            date [total, acc, risk],
            ... (from after to before)
            until N dates
          ]
          ...
          until mangers number.
        ]
  **/
  constructor (wg, data) {
    this._wg = wg;
    this._data = data;
    this._mSel = -1;

    this.view();
  }

  // View ----------------------------------------------------------------------

  /**
      @private
  **/
  view () {
    const wg = $("div");
    const data = this._data;
    const mSel = this._mSel === -1 ? "All" : _("Inv-") + String(this._mSel);

    const lopts = [
      Menu.toption("All", _("All"), () => this.setMenu(-1))
    ];
    for (let i = 0; i < data.length; ++i) {
      const op = _("Inv-") + String(i);
      lopts.push(Menu.separator());
      lopts.push(Menu.toption(op, op, () => this.setMenu(i)));
    }
    const ropts = [];
    const menu = new Menu(lopts, ropts, mSel);

    let d = [];
    if (this._mSel === -1) {
      for (let i = 0; i < data[0].length; ++i) {
        let total = 0;
        let acc = 0;
        let risk = 0;
        for (let inv = 0; inv < data.length; ++inv) {
          if (data[inv][i]) {
            const values = data[inv][i][1];
            total += values[0];
            acc += values[1];
            risk += values[2];
          }
        }
        d.push([data[0][i][0], [total, acc, risk]]);
      }
    } else {
      d = data[this._mSel];
    }
    ProfitsWg.mk(wg, d);

    this._wg
      .removeAll()
      .add(menu.wg)
      .add(wg)
    ;
  }

  // Control -------------------------------------------------------------------

  /**
      @private
      @param {number} manager
      @return void
  **/
  setMenu (manager) {
    this._mSel = manager;
    this.view();
  }

  // Static --------------------------------------------------------------------

  /**
      @param {!Domo} wg
      @return !Promise<!Profits>
  **/
  static async mk (wg) {
    const rp = await Cts.client.send({
      "module": "acc",
      "source": "profits",
      "rq": "idata"
    });

    const data = rp["data"];
    return new Profits(wg, data);
  }

}

File: app/MultiMarket/go/data/profits/profits.go
// Copyright 17-Sep-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// Profits historic entry.
package profits

import (
	"github.com/dedeme/golib/json"
)

type T struct {
	date  string
	total float64
	acc   float64
	risk  float64
}

// Create an entry
//   date : Date in base format (YYYYMMDD)
//   total: Total profits.
//   acc  : Accounting profits.
//   risk : Risk profits.
func New(date string, total, acc, risk float64) *T {
	return &T{date, total, acc, risk}
}

// Date in base format (YYYYMMDD)
func (e *T) Date() string {
	return e.date
}

// Total profits.
func (e *T) Total() float64 {
	return e.total
}

// Accouning profits.
func (e *T) Acc() float64 {
	return e.acc
}

// Risk profits.
func (e *T) Risk() float64 {
	return e.risk
}

func (e *T) ToJs() json.T {
	return json.Wa([]json.T{
		json.Ws(e.date),
		json.Wd(e.total),
		json.Wd(e.acc),
		json.Wd(e.risk),
	})
}

func FromJs(js json.T) *T {
	a := js.Ra()
	return &T{
		a[0].Rs(),
		a[1].Rd(),
		a[2].Rd(),
		a[3].Rd(),
	}
}
